Add API Gateway IAM (AWS_IAM) Authorizers

horike37 opened the pull request:
*Issue #802 #781,  if available:*

*Description of changes:*
Add API Gateway IAM (AWS_IAM) Authorizers

*Description of how you validated changes:*
Please verify this PR with following the RFC of #781,

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation
Regarding documentation, I didn't update those because I took a look at stuff under `doc` directory but I was not able to find related to `Authorizers` features. Ping me if missed it.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

brettstack commented on tests/translator/input/api_with_aws_iam_auth_overrides.yaml:
```diff
@@ -0,0 +1,54 @@
+            RestApiId: !Ref MyApiWithAwsIamAuth
+            Method: get
+            Path: /2
+  MyFunction3:
```
Edited the above comment. I don't think NONE should be a valid value for Invoke Role.

horike37 commented on samtranslator/model/api/api_generator.py:
```diff
@@ -248,14 +248,23 @@ def _add_auth(self):
+            authorizers[default_authorizer] = ApiGatewayAuthorizer(
+                api_logical_id=self.logical_id,
+                name=default_authorizer,
+                default_authorizer=default_authorizer
```
make sense.

horike37 commented on samtranslator/swagger/swagger.py:
```diff
@@ -104,9 +104,30 @@ def add_lambda_integration(self, path, method, integration_uri):
                 'uri': integration_uri
             }
 
+        if api_auth_config is not None:
```
I think so, will do

horike37 commented:
Hey @brettstack, thank you for the review 👍 
I have updated all of them. Please check it again.

brettstack commented:
We'd also appreciate an example using AWS_IAM. If you're not up for it, we'll ask someone else if they'd like to have contribute one. See this example as a reference https://github.com/awslabs/serverless-application-model/tree/master/examples/2016-10-31/api_lambda_token_auth.

keetonian commented:
Looks like I broke things with my last merge. I'll fix that.

horike37 commented:
Hey, @brettstack @brettstack. I have fixed the conflicts and added the documentation and source comment. 

>We'd also appreciate an example using AWS_IAM. If you're not up for it, we'll ask someone else if >they'd like to have contribute one. See this example as a reference >https://github.com/awslabs/serverless-application-model/tree/master/examples/2016-10-31/api_lambda_token_auth.

Sorry, it is difficult for me to spend more time to do it, would be great if you can ask someone else.

horike37 commented:
Thanks @brettstack @keetonian 😄 
When will this be released? This is needed for for my job.

brettstack commented:
Policy says we can't give dates, but I can say we want to get this and other features/fixes out asap.

If you need it now, you can try using SAM CLI and these instructions for using a dev version of SAM https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#4-optional-install-development-version-of-sam-transformer. Running `sam validate --debug` you'll get the transformed template.

Jun711 commented:
@brettstack @horike37 
While waiting for this pull request to be released, would using a securityDefinitions work? thanks
```
securityDefinitions: 
   AwsIamAuthorizer:
      type: "apiKey"
      name: "Authorization"
      in: "header"
      x-amazon-apigateway-authtype: "awsSigv4"
```
and, under my AWS::Serverless::Api path method, I add a security item that points to the securityDefinition.
```
 security: 
 - AwsIamAuthorizer: []
```
